packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,221 @@
+  }
+  if (
+    actualNode &&
+    expectedNode &&
+    typeof actualNode == 'object' &&
+    typeof expectedNode == 'object'
+  ) {
+    if (Array.isArray(actualNode) && Array.isArray(expectedNode)) {
+      if (actualNode.length != expectedNode.length) {
+        return false;
+      }
+      return !actualNode.some(
+        (nodeEle, index) => !astNodesAreEquals(nodeEle, expectedNode[index]),
+      );
+    }
+    if (isAstNodeType(actualNode) && isAstNodeType(expectedNode)) {
+      const actualNodeKeys = Object.keys(actualNode).filter(
+        key => !astIgnoreKeys.includes(key),
+      );
+      const expectedNodeKeys = Object.keys(expectedNode).filter(
+        key => !astIgnoreKeys.includes(key),
+      );
+      if (actualNodeKeys.length === expectedNodeKeys.length) {
+        return !actualNodeKeys.some(
+          actualNodeKey =>
+            !astNodesAreEquals(
+              actualNode[actualNodeKey],
+              expectedNode[actualNodeKey],
+            ),
+        );
+      }
+    }
+  }
+  return false;
+};
+
+const isAstNodeType = (node: object): node is Record<string, TSESTree.Node> => {
+  return node.constructor === Object;
+};
+
+export type Options = [
+  {
+    ignoreIntersections?: boolean;
+    ignoreUnions?: boolean;
+  },
+];
+
+export type MessageIds = 'duplicate' | 'suggestFix';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-duplicate-type-constituents',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow duplicate union/intersection type members',
+      recommended: false,
+    },
+    fixable: 'code',
+    hasSuggestions: true,
+    messages: {
+      duplicate: '{{type}} type member {{name}} is duplicated.',
+      suggestFix: 'Delete duplicated members of type (removes all comments).',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreIntersections: {
+            type: 'boolean',
+          },
+          ignoreUnions: {
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreIntersections: false,
+      ignoreUnions: false,
+    },
+  ],
+  create(context, [{ ignoreIntersections, ignoreUnions }]) {
+    const sourceCode = context.getSourceCode();
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const duplicateConstituents: {
+        node: TSESTree.TypeNode;
+        index: number;
+      }[] = [];
+
+      node.types.reduce<TSESTree.TypeNode[]>((acc, cur, index) => {
+        if (acc.some(ele => constituentsAreEqual(ele, cur))) {
+          duplicateConstituents.push({ node: cur, index });
+          return acc;
+        }
+        return [...acc, cur];
+      }, []);
+
+      const hasComments = node.types.some(type => {
+        const count =
+          sourceCode.getCommentsBefore(type).length +
+          sourceCode.getCommentsAfter(type).length;
+        return count > 0;
```

Comment:
Nit: if `getCommentsBefore` has entries, we don't need to call `getCommentsAfter`.

```suggestion
        return (
          sourceCode.getCommentsBefore(type).length ||
          sourceCode.getCommentsAfter(type).length
        );
```